app/src/main/java/org/mozilla/tv/firefox/webrender/cursor/CursorModel.kt

```diff
@@ -0,0 +1,352 @@
+                lastUpdatedAtMS = currTime
+                return true
+            }
+        }
+    }
+
+    private fun calculateAndSendScrollEvent(timePassed: Long, velocity: Float, oldPos: PointF) {
+        // This should not live inside internalMutatePositionAndReturnVelocity.
+        // Move it if you can find a better solution
+        val approxFramesPassed = (timePassed / 16).toInt()
+        val shouldScroll = getScrollDistance(velocity, oldPos, approxFramesPassed)
+        shouldScroll?.let { _scrollRequests.onNext(it) }
+    }
+
+    /**
+     * Mutates [oldPos] to its new position. This position is calculated based on the time passed
+     * since the last update, in order to avoid miscalculations when dropping frames.
+     *
+     * Calculates necessary scrolling, if any, and sends out a request to handle it.
+     *
+     * @returns the new cursor velocity
+     */
+    private fun internalMutatePositionAndReturnVelocity(
+        oldPos: PointF,
+        timePassedSinceLastMutation: Long,
+        oldVelocity: Float,
+        directionsPressed: Set<Direction>
+    ): Float {
+        // directionsPressed empty case is handled in `mutatePosition`
+        require(directionsPressed.isNotEmpty())
+        val screenBounds = screenBounds ?: return 0f
+
+        val accelerateBy = ACCELERATION_PER_MS * timePassedSinceLastMutation
+        val velocity = (oldVelocity + accelerateBy).coerceIn(0f, MAX_VELOCITY)
+
+        fun updatePosition() {
+            var verticalVelocity = 0f
+            if (directionKeysPressed.contains(Direction.UP)) verticalVelocity -= velocity
+            if (directionKeysPressed.contains(Direction.DOWN)) verticalVelocity += velocity
+            var horizontalVelocity = 0f
+            if (directionKeysPressed.contains(Direction.LEFT)) horizontalVelocity -= velocity
+            if (directionKeysPressed.contains(Direction.RIGHT)) horizontalVelocity += velocity
+            oldPos.x += horizontalVelocity
+            oldPos.y += verticalVelocity
+            oldPos.x = oldPos.x.coerceIn(0f, screenBounds.x)
+            oldPos.y = oldPos.y.coerceIn(0f, screenBounds.y)
+        }
+
+        updatePosition()
+
+        return velocity
+    }
+
+    /**
+     * If the cursor is near the edge of the screen, this will return that
+     * direction.  If not, it will return null.
+     *
+     * When in a corner, only UP or DOWN will be returned.
+     */
+    fun getEdgeOfScreenNearCursor(): Direction? {
+        val screenBounds = screenBounds ?: return null
+        return when {
+            lastKnownCursorPos.y < EDGE_OF_SCREEN_MARGIN -> Direction.UP
+            lastKnownCursorPos.y > screenBounds.y - EDGE_OF_SCREEN_MARGIN -> Direction.DOWN
+            lastKnownCursorPos.x < EDGE_OF_SCREEN_MARGIN -> Direction.LEFT
+            lastKnownCursorPos.x > screenBounds.x - EDGE_OF_SCREEN_MARGIN -> Direction.RIGHT
+            else -> null
+        }
+    }
+
+    private fun resetCursorSpeed() {
```

Comment:
nit: resetCursorState

Since this also clears the direction keys, this does more than resetCursor position (but clearing keys is only used in one place: maybe we should move that?)